### PR TITLE
Make this work on ci.j.io

### PR DESCRIPTION
**Pull request by daniel-beck**
Work in progress

/me sighs

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -73,6 +73,10 @@
 
   <repositories>
     <repository>
+      <id>maven.jenkins-ci.org</id>
+      <url>https://repo.jenkins-ci.org/public/</url>
+    </repository>
+    <repository>
       <id>m.g.o-public</id>
```
Comment (jglick) [inline pom.xml]: Delete this.

File: maven-settings.xml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<settings>
+  <mirrors>
+    <mirror>
+      <id>repo.jenkins-ci.org</id>
+      <url>https://repo.jenkins-ci.org/public/</url>
```
Comment (jglick) [inline maven-settings.xml]: Why is this here? The `pom.xml` already specifies this repository.

Comment (daniel-beck) [inline maven-settings.xml]: This is used in a separate `mvn` invocation that downloads dependencies of core and all plugins before scanning for extension implementations (perhaps to make it compilable, unclear to me).

My current best guess is that without this, the standalone `mvn` invocation trying to download `jenkins-core` dependencies doesn't find the parent pom that's not in the parent dir.

File: src/main/java/org/jenkinsci/extension_indexer/ExtensionPointListGenerator.java [hunk 3]
```diff
@@ -304,7 +304,7 @@ private MavenRepositoryImpl createRepository() throws Exception {
      * Walks over the plugins, record {@link #modules} and call {@link #discover(Module)}.
      */
     private void processPlugins(MavenRepository r) throws Exception {
-        ExecutorService svc = Executors.newFixedThreadPool(4);
+        ExecutorService svc = Executors.newFixedThreadPool(1);
```
Comment (kwhetstone) [inline src/main/java/org/jenkinsci/extension_indexer/ExtensionPointListGenerator.java]: Does this drastically increase the runtime?

Comment (daniel-beck) [inline src/main/java/org/jenkinsci/extension_indexer/ExtensionPointListGenerator.java]: Yes and no.

Yes, because things go slower (while they're still going).

No, because it kept locking up on me with stack traces like https://gist.github.com/daniel-beck/401a85db7d58a55b0c78e02d33f971fd

Comment (daniel-beck) [inline src/main/java/org/jenkinsci/extension_indexer/ExtensionPointListGenerator.java]: FWIW I can try reverting this one the PR build passes to see whether that can finish with 4 threads. If so, I won't care about local execution.

File: maven-settings.xml (hunk 2 above)
Comment (daniel-beck) [inline maven-settings.xml]: Additionally, plugins last released before 2012 or so will reference now obsolete Maven repositories.

File: src/main/java/org/jenkinsci/extension_indexer/SourceAndLibs.java [hunk 4]
```diff
@@ -143,6 +143,8 @@ private static void downloadDependencies(File pomDir, File destDir) throws IOExc
             process = System.getenv("M2_HOME") + "/bin/mvn";
         }
         ProcessBuilder builder = new ProcessBuilder(process,
+                "--settings", new File("maven-settings.xml").getAbsolutePath(),
```
Comment (jglick) [inline src/main/java/org/jenkinsci/extension_indexer/SourceAndLibs.java]: BTW it is poor style to assume that the CWD of the process has a particular file in it. Rather out `maven-settings.xml` into `src/main/resources/` and use `FileUtils` to copy the resource to a local temp file.

File: src/main/java/org/jenkinsci/extension_indexer/SourceAndLibs.java [hunk 5]
```diff
@@ -143,6 +143,8 @@ private static void downloadDependencies(File pomDir, File destDir) throws IOExc
...
+                "--update-snapshots",
                 "dependency:copy-dependencies",
```
Comment (jglick) [inline src/main/java/org/jenkinsci/extension_indexer/SourceAndLibs.java]: Maybe try running the `validate` phase first:

    mvn -s … -U -DincludeScope=compile -DoutputDirectory=… validate dependency:copy-dependencies

**Comment by daniel-beck** [comment]
The previous PR build got to `mber`, this one is stuck at `notification`. Not a huge difference.

**Comment by daniel-beck** [comment]
Re-running the PR so I can check whether it's again disk usage.

**Comment by daniel-beck** [comment]
Node monitoring showed that disk space was at just over 4GB when I aborted the build just now (after ~1.5hrs), so it's safe to assume it'd just run out.

**Comment by daniel-beck** [comment]
Related: https://issues.jenkins-ci.org/browse/INFRA-1190

**Comment by daniel-beck** [comment]
@rtyler did some hackery on the highmem agent and currently it reports 362.53 GB free disk space. This should be enough for this build 😃

**Comment by daniel-beck** [comment]
It's at 'm', with 348.01 GB free disk space.

**Comment by daniel-beck** [comment]
After 3hr 18min it's at 'toolenv', with 340.61 GB free disk space.

**Comment by daniel-beck** [comment]
3 hours 38 minutes. 🎉

File: src/main/java/org/jenkinsci/extension_indexer/ExtensionPointListGenerator.java (hunk 3 above)
Comment (timja) [reply src/main/java/org/jenkinsci/extension_indexer/ExtensionPointListGenerator.java:307]: Appears to have been fixed in https://github.com/jenkins-infra/backend-extension-indexer/pull/35 by ripping out the affected code.

Testing locally with a patch using increased threads
